feat: add client based logger

kenkalang opened the pull request:
See https://github.com/eclipse-paho/paho.mqtt.golang/issues/503

Changes:
- Introduced clientLogger to encapsulate loggers per client.
- Added ClientOptions methods to allow setting loggers for each level (Error, Critical, Warn, Debug), all scoped to the ClientID.
- Replaced global logging calls with client-bound clientLogger usage.
- Injected clientLogger into relevant functions and structs to maintain client-level context in logs.

Before:
```Go
mqtt.ERROR = log.New(os.Stdout, "[ERROR] ", 0)
mqtt.CRITICAL = log.New(os.Stdout, "[CRITICAL] ", 0)
mqtt.WARN = log.New(os.Stdout, "[WARN]  ", 0)
mqtt.DEBUG = log.New(os.Stdout, "[DEBUG] ", 0)
```

After:
```Go
errorLogger := log.New(os.Stdout, "[ERROR]", 0)
criticalLogger := log.New(os.Stdout, "[CRITICAL]", 0)
warnLogger := log.New(os.Stdout, "[WARN]", 0)
debugLogger := log.New(os.Stdout, "[DEBUG]", 0)

opts = mqtt.NewClientOptions().
	SetClientID(clientID).
	SetDebugLogger(debugLogger).
	SetErrorLogger(errorLogger).
	SetCriticalLogger(criticalLogger).
	SetWarnLogger(warnLogger)
```

Example output: 
<img width="423" alt="Screenshot 2025-07-04 at 18 35 21" src="https://github.com/user-attachments/assets/54ae412e-e425-4b37-99a1-b9832e23281e" />

Note:

This log usage and output is no longer relevant. I changed the client log to use sLog, as for now there will be 2 log, global and client. See latest commit for more details

MattBrittan commented:
Thanks very much for all of your effort on this.  But please consider in commenting on the issue before putting too much effort into changes (can be difficult to get much of a response, but it's worth trying). 

Note: I do not personally use this library any more (have moved everything to V5) so will make a comment on the original PR asking for input from others (this also limits the amount of time I spend on this library). This also means that these comments are based on a fairly quick look (but with the extent of these changes a quick look is not all that quick!).

As I noted in issue #503 "My preference would be that the client defaults to the global logger because this would avoid breaking existing changes". Go has a bit of a focus on stability (see the [[Go 1 compatibility guarantee](https://go.dev/doc/go1compat)](https://go.dev/doc/go1compat)) and this applies to libraries (more so than a lot of other languages). Because of this (and experiences with other languages!) I've tried very hard to avoid breaking changes in this library (this is why the V5 client is completly seperate). As your PR containa breaking changes, we need to either:

* Move to a new major version ("V2") because users would expect a minor version change to be non-breaking; or
* Remove the breaking changes

My preference would be option 2 as I don't think this would be all that difficult. Based on a quick look:
- `NewClientOptions()` - Default to logger that utilises the global loggers defined in `trace.go` (meaning as long as users set the loggers before calling `NewClientOptions()` logs will operate as they have previously)
- `MQTT.NewFileStore` function definition changed (also with other stores) - change this to use default loggers and add a new `NewFileStoreEx`, or similar, to take the new logger.
- `ConnectionLostHandler` signature change - users can get the logger from the `Client` (already passed).

A few other general thoughts:
- `func NewClientLogger(clientID string, errorLogger, criticalLogger, warnLogger, debugLogger Logger) *clientLogger `, this may be difficult to use because `clientLogger` is not exported (so a user cannot do `var *clientLogger paho.clientLogger`). I also wonder whether using an `interface` for `SetErrorLogger` etc would be better (means a user can implement their own logger (something you can currently do but would no longer be possible with this PR). 
- Consider removing all direct calls to the old `trace.go` functions - there are still calls in `options.go` and `store.go`.
- If making a change this big is it worth consideting moving to `sLog`. I believe the main benefir from this is that the log can more easily filtered etc. This might be too big of a change.

kenkalang commented:
Hi Matt, thanks a lot for your feedback, I really appreciate the time you've taken, especially since you're no longer actively using this library.
You're right that I should have opened a discussion first before jumping into the code. I’ll make sure to do that in the future to avoid unnecessary work and ensure alignment before making larger changes.

- I’ve adjusted the code to **remove all breaking changes**, as suggested.
- For `DefaultConnectionLostHandler`, I reverted to using the global `DEBUG` logger directly. Since `Client` is an interface, introducing a new method like `Logger()` would have been a breaking change, so I avoided that.
- In `options.go`, there’s still a direct call to the global logger (`AddBroker`), unfortunately this occurs before a client instance is created, so client-based logging isn’t possible at that stage.

Do you have any suggestions regarding my last 2 points?

Lastly, can you elaborate more on how adding sLog could address logging of message payload issues?

Thanks again for your feedback

kenkalang commented:
There are some changes:
- Client log uses sLog
- Add back global log

There are 2 logs for now, client and global, just to avoid any breaking changes. Will remove the global log soon

dhruvjain99 commented on options.go:
```diff
@@ -105,17 +105,19 @@ type ClientOptions struct {
+//	KeepAlive: 30 (seconds)
+//	ConnectTimeout: 30 (seconds)
+//	MaxReconnectInterval 10 (minutes)
+//	AutoReconnect: True
```
Nit: We can add LogVerbosity default here

kenkalang commented on client.go:
```diff
@@ -165,9 +165,20 @@ func NewClient(o *ClientOptions) Client {
 	c.persist = c.options.Store
-	c.messageIds = messageIds{index: make(map[uint16]tokenCompletor)}
-	c.msgRouter = newRouter()
+	c.messageIds = messageIds{index: make(map[uint16]tokenCompletor), logger: *c.logger}
```
Thanks for pointing that out. I initially passed slog.Logger by value, thinking it was a lightweight struct. But then I realized that it holds internal state. I'll change it

kenkalang commented on net.go:
```diff
@@ -40,64 +41,75 @@ const closedNetConnErrorText = "use of closed network connection" // error strin
 //
 // Note that, for backward compatibility, ConnectMQTT() suppresses the actual connection error (compare to connectMQTT()).
 func ConnectMQTT(conn net.Conn, cm *packets.ConnectPacket, protocolVersion uint) (byte, bool) {
```
You're right, adding another fn such `ConnectMQTTEx()` makes more sense

dhruvjain99 commented on options.go:
```diff
@@ -149,6 +151,7 @@ func NewClientOptions() *ClientOptions {
 		Dialer:                  &net.Dialer{Timeout: 30 * time.Second},
 		CustomOpenConnectionFn:  nil,
 		AutoAckDisabled:         false,
+		LogVerbosity:            LogLevelDefault,
```
Yes, it is fine with mqtt.LogLevel

MattBrittan commented:
Sorry - had planned on looking at this before the end of the week but thats not going to happen. Will aim for next week (this is quite a big change now so takes time to review).

MattBrittan commented on options.go:
```diff
@@ -450,8 +453,15 @@ func (o *ClientOptions) SetCustomOpenConnectionFn(customOpenConnectionFn OpenCon
 }
+
+// SetLogLevel sets the log level for the client. This will be used to control the verbosity of the logs
+func (o *ClientOptions) SetLogLevel(logVerbosity LogLevel) *ClientOptions {
```
As commented elsewhere I think `SetLogger(l *slog.Logger) *ClientOptions` might be more flexible? 

A few reasons:
* One of the main belefits of using `slog` is that it's part of the standard library and there are heaps of adaptors (for example I use `zerolog` in quite a few apps and could use `https://github.com/samber/slog-zerolog` to get an `slog.Logger` to pass into such a function).
* Users can use whatever output type they want

MattBrittan commented on client.go:
```diff
@@ -386,22 +424,28 @@ func (c *client) attemptConnection() (net.Conn, byte, bool, error) {
 			ERROR.Println(CLI, err.Error())
+			c.logger.Error("Failed to connect to broker", slog.String("error", err.Error()), componentAttr(CLI))
 			WARN.Println(CLI, "failed to connect to broker, trying next")
+			c.logger.Warn("failed to connect to broker, trying next", componentAttr(CLI))
```
Is this needed when there is an Error for the same thing? (think we want one or the other, not both)

MattBrittan reviewed (commented):
Confirmed that pre PR CMD files all build OK and pre PR tests (with changes to private fn sigs only) pass (there is one dodgy test that fails intermittently prior to this PR). This reassures me that there should not be breaking changes.

Thanks very much for your work on this. I think it will be a worthy change but would prefer that we don't end up with two logging engines if thats possible (have made some suggestions as to how this could be accomplished). Note: my comments are suggestions - please feel free to push back!

As this is a big change so have created issue #722  to try and involve more people. Think I'l also do a dependency update/release this weekend so we can add this PR into Master and let it sit a while before it gets into a release.

kenkalang commented on trace.go:
```diff
@@ -42,3 +47,14 @@ var (
 	DEBUG    Logger = NOOPLogger{}
 )
+
+var noopSLogger = slog.New(slog.NewTextHandler(io.Discard, &slog.HandlerOptions{}))
```
In that case we’d also need to upgrade the Go version to 1.24, since slog.DiscardHandler was only added in 1.24.

MattBrittan replied:
Did not realise that :-). Upgraging to 1.24 is fine (policy is to try to match supported Go Versions, so stay one version behind current release).  However it's probably not worth it for this minor change.

kenkalang commented:
Hi @MattBrittan, I’ve committed a change reflecting your suggestion, please do take a look and give it a review. One thing to note is that I’ve mapped the `CRITICAL` log to Warn rather than creating a separate custom handler. Keeping everything within slog’s model avoids us having to maintain yet another handler abstraction, and helps push us toward deprecating the older style calls cleanly in the near future.

Other than that, there are still a couple of places I haven’t resolved yet:

- In `Options.go`, there’s an `ERROR` call in `AddBroker` before the client logger has been initialized. In that case, I think it might be cleaner to return an error instead of logging.
- In `DefaultConnectionLostHandler`, we can likely remove the logging entirely since afaik it’s only a placeholder.

Once we agree on this, we should be able to simplify further and remove the old calls entirely. I’m open to any suggestions here, so please let me know your opinion on this.

MattBrittan commented:
As mentioned above I was keen to make a release before looking at this further (as thie is a large change), I have now done that with 1.5.1 (apologies - this introduced a few conflicts). I aim to have a look at the logging over the next couple of weeks (sorry have put quite a bit of time into the releases and need to catch up on other stuff!).

MattBrittan commented:
- Copied main library changes across to current master and confirmed that all tests pass and all examples build (confirming that the changes don't break existing examples)
- Same setup - confirmed that docker example runs and log entries make sense (with loggine fully activated)
- Reviewed changes again - all look OK so will accept (probably missed a thing or two but don't want to leave this large PR hanging around longer)

MattBrittan commented:
@kenkalang Thanks very much for all of your work on this, and sorry it's taken so long to accept.
